PEP 809: Stable ABI for the Future (#4599)
python/peps: peps/pep-0809.rst

Related PEPs
============

:pep:`803` is an alternative to this proposal, and much of the background text
used in this PEP is deliberately identical. We thank the PEP 803 author for the
use of their text.

Background
----------

Python's Stable ABI, as defined in :pep:`384` and :pep:`652`, provides a way to
compile extension modules that can be loaded on multiple minor versions of the
CPython interpreter.
Several projects use this to limit the number of
:ref:`wheels <packaging:binary-distribution-format>` (binary artefacts)
that need to be built and distributed for each release, and/or to make it
easier to test with pre-release versions of Python.

With free-threading builds (:pep:`703`) being on track to eventually become
the default (:pep:`779`), we need a way to make the Stable ABI available
to those builds.

To build against the Stable ABI, the extension must use a *Limited API*,
that is, only a subset of the functions, structures, etc. that CPython
exposes.
The Limited API is versioned, and building against Limited API 3.X
yields an extension that is ABI-compatible with CPython 3.X and *any* later
version (though bugs in CPython sometimes cause incompatibilities in practice).

> Also, the Limited API is not "stable": newer versions may remove API that
> were a part of older versions.

Reviewer: ```suggestion
Also, the Limited API is not "stable": newer versions may remove API items that
were available in older versions.
```

> This PEP proposes a significant change to versioning of both the Limited API
> and the Stable ABI, with the goal of enabling long-term management of stability
> and compatibility, but without preventing access to innovations for those who
> are using the limited subsets.

Reviewer: ```suggestion
This PEP proposes a significant change to the versioning of the Limited API
and the Stable ABI. The goal is to enable long-term management of stability
and compatibility, while also allowing access to innovations when using the limited subsets.
```

   for the 3.14 free-threaded interpreter, and for older stable ABI versions.

API changes are OK
   The new Limited API may require extension authors to make significant
   changes to their code.
   Projects that cannot do this (yet) can continue using Limited API 3.14,
   which will yield extensions compatible with GIL-enabled builds only.

No extra configuration
   We do not introduce new "knobs" that influence what API is available
   and what the ABI is compatible with.

Specification
=============

Note that much of the specification is identical to :pep:`803`, and readers
should refer to that proposal for details. The ABI Stability, Build-time Macros
and Interfaces API sections are unique to this proposal.

ABI Stability
-------------

The Stable ABI will be frozen for a duration of at least 10 years. When a new
version of the Stable ABI is frozen, the existing version will continue to be
supported for at least 5 years. This allows ample migration time for consumers
(that is, package developers) to migrate their entire range of supported
releases simultaneously. However, if the core team sees no reason to replace the

> current Stable ABI, the new version may be deferred.

Reviewer: Which core team: CPython or package maintainer?

PEP: 809
Title: Stable ABI for the Future
Author: Steve Dower <steve.dower@python.org>
Discussions-To: Pending
Status: Draft
Type: Standards Track
Created: 19-Sep-2025

> Python-Version: 3.15

Author: Added the Requires, but I'll get it merged and then add the post history for the real post, not the hijacked one.

Abstract
========

The Stable ABI as ``abi3`` can no longer be preserved, and requires replacement.
``abi2025`` will be the first replacement, providing resolution of current known
incompatibilities, with planned retirement after 10 years. A new ABI ``abi2030``
is planned to provide five years of overlap for the next ABI.

Long-term stability will be enabled through a mechanism for runtime ABI
discovery, allowing extensions to be run with earlier releases that support the
same ABI. Changes and additions during the lifespan of an ABI can be added as
*interfaces*, allowing them to be discovered at runtime so that callers can
choose suitable fallback behaviour. Currently, such additions prevent extensions
from loading at all on earlier runtimes.

> The ``abi3`` ABI will be retained in GIL-enabled builds for five years, after

Author: Clarified - it _might_ be retired, but that only really matters if we decide to abandon free-threaded and keep the GIL-enabled builds. Otherwise I expect it'll remain until those get retired (but I'm also not going to enforce that in this PEP, hence being a bit vague)

``abi3`` will select the "most complete" version of ABI3 available in that
release. For example, using ``0x03ff2026`` in 3.15+ would select ``abi2026``,
while in 3.10 will select the version of ABI3 that works for 3.10-3.14.

Wheel tags
----------

Wheels should be tagged with the ABI tag ``abi2026``. No changes to Python or
platform tags are needed. It is perhaps worth noting that releases tagged for
``cp314`` or earlier will never be compatible with ``abi2026``, as it was not
present, and so a wheel tagged ``py3-abi2026-<plat>`` is not going to cause a
wheel using the new Stable ABI to be loaded by an older release.

New API
-------

Implementing this PEP will make it possible to build extensions that
can be successfully loaded on free-threaded Python, but not necessarily ones
that are thread-safe without a GIL.

Limited API to allow thread-safety without a GIL -- presumably ``PyMutex``,
``PyCriticalSection``, and similar -- will be added via the C API working group,
or in a follow-up PEP.

Interfaces API
--------------

A new interfaces API will be added to Python and the new Limited API. This API
is to satisfy the "semantic changes are detectable on all releases" requirement

> from the ABI Stability section above. That is, consumers will be able to adopt

Author: Everyone seems to think this is only about extension developers and not all consumers of the API... is there a syntax that will let me add a footnote here to clarify?

maintainers to immediately support each new ABI.

Forward-and-backward compatibility is ensured by dynamic interface detection.
Code using recently added limited API functions will run on older releases,
though potentially at lower performance. See the documentation for new functions
to find information about any Limited API-specific nuances.

Non-C callers should use the interfaces mechanism directly to get access to new
features without artificially limiting their compatibility to newer releases.
The names and struct layouts of interfaces are guaranteed stable for all time,
though it should not be assumed that an interface will be available for all
time, and suitable fallback code (either an alternative implementation or error
handling) should be included.

Reference Implementation
========================

See :pep:`803` for links to reference implementations for the aspects inherited
from that PEP.

The reference implementation of interfaces is
`zooba/cpython#44 <https://github.com/zooba/cpython/pull/44/files>`__.

Rejected Ideas
==============

[See discussion for now.]

Open Issues
===========

[See discussion for now.]

Footnotes

> =========

Author: I thought you were suggesting I should add those two lines and was very confused... 😆
